### PR TITLE
[SPARK-32850][CORE][K8S] Simplify the RPC message flow of decommission

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

This PR cleans up the RPC message flow among the multiple decommission use cases, it includes changes:

* Keep `Worker`'s decommission status be consistent between the case where decommission starts from `Worker` and the case where decommission starts from the `MasterWebUI`: sending `DecommissionWorker` from `Master` to `Worker` in the latter case.

* Change from two-way communication to one-way communication when notifying decommission between driver and executor: it's obviously unnecessary for the executor to acknowledge the decommission status to the driver since the decommission request is from the driver. And it's same in reverse.

* Only send one message instead of two(`DecommissionSelf`/`DecommissionBlockManager`) when decommission the executor: executor and `BlockManager` are in the same JVM.

* Clean up codes around here.

### Why are the changes needed?

Before:

WeChat56c00cc34d9785a67a544dca036d49da

After:
WeChat05f7afb017e3f0132394c5e54245e49e

(Note the diagrams only counts those RPC calls that needed to go through the network. Local RPC calls are not counted here.)

After this change, We reduced 6 original RPC calls and added one more RPC call for keeping the consistent decommission status for the Worker. And the RPC flow becomes more clear.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Updated existing tests.

### CODE REVIEW

**Comment by Ngone51** [comment]
Bring this back since it isn't the original commit that breaks the K8S test. As the PR https://github.com/apache/spark/pull/29751 merged before this already failed the K8S tests

**Comment by HyukjinKwon** [comment]
@holdenk, seems like the test failure wasn't caused by this PR. Dose your -1 at [here](https://issues.apache.org/jira/browse/SPARK-32937?focusedCommentId=17198867&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17198867) still stand?

cc @dongjoon-hyun as well per #29408.

**Comment by holdenk** [comment]
The -1 is only on the PR that I mentioned it on (the refactoring with many sub classes) which still stands. I would like to review this PR more though since I think we probably need better test coverage of this change than it originally had. Sound good? Thanks for checking in about that :)

**Comment by HyukjinKwon** [comment]
Got it about -1 but what about we pushing this as is, and work on the test as followups? It's a bit odds that we reverted it for the reason this PR didn't cause, and ask more things before merging it in.

**Comment by holdenk** [comment]
I think we should not commit this with the K8s test being broken. It’s in the same chunk of code and changes the logging string (although another PR also changed that string first too?). I do not believe this PR was appropriately tested when first merger given it changed decommissioning messages and did not run the decommission tests.

For clarity: if you want to fix the tests in a separate PR that’s ok with me, but I would prefer not to commit this without passing integration testing.

**Comment by Ngone51** [comment]
> I think we should not commit this with the K8s test being broken.

We don't. That's also why I added `[K8S]` tag in the PR title. And feel free to leave comments, I can address them in followups.

**Comment by HyukjinKwon** [comment]
@holdenk, why don't you take a look for a test failure since it blocks all changes in decommission in k8s, and you were involved mainly in the development there?

**Comment by cloud-fan** [comment]
Can we find out which commit caused the test failure in the first place? We should either revert that commit, or fix it soon, as the test failure blocks others.

Since this PR is resubmitted (although the revert is not necessary now given the test failure was already there), I think it's a good chance for @holdenk to take a closer look before re-merging. And I agree with @holdenk that we can't merge a PR when the related test is already broken. We should fix that first. @holdenk can you give some hints about it? I took a look at https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33556/ , but I don't even see how the test failed. The output is very different from normal Spark tests.

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:74]: I think this might mean we return from handling the signal right away rather than waiting for decommissionSelf to be finished. Is this an intentional change?

Also this will no longer report a decommissioning failure by signal return value, so may block pod deletion or other cleanup task longer than needed.

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:680]: May I suggest "Tell the master that we are starting decommissioning so it stops trying to launch executor/driver on us"

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:92]: Similar comments as in Worker.scala

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:179]: I don't think we should just take this out. async sends could fail, re-sending the message if we receive a request which indicates the master hasn't received our notification indicates we should resend.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:216]: s/we starts/we are starting/
?

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:215]: Doing this after decommissionSelf introduces a race condition, not too important one but given that it wasn't in the original flow I'd suggest swapping the order.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:272]: Here might be where you break the test suite last time, so double check it.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala:1812]: Why?

**Review by holdenk — COMMENTED** [review]
Thanks for reverting this and resubmitting it. I know you believe the original PR didn't cause the test failure, but that's only half true, this PR just broke the test some more.

That being said I still have concerns this PR is not sufficiently tested, can you add some more tests for the new flows you've introduced?

**Comment by holdenk** [comment]
I really like the idea of simplifying the RPC message flow, thanks for taking this on @Ngone51 and I'm sorry the code here is so brittle to these types of changes (the K8s integration tests are kind of limited).

**Comment by HyukjinKwon** [comment]
> but that's only half true, this PR just broke the test some more.

@holdenk, you're kidding right? There was only one test failure that was not caused by this PR in K8S tests. That test was just fixed by you. How come this PR broke more tests? Can you be more explicit on that? Which tests were more broken, and how did you test?

**Comment by holdenk** [comment]
I’m not joking. The test has multiple conditions. Another PR broke one of the conditions. This PR in its original broke another one of the conditions. It was reported in the same test because K8s integration tests focus on the integration so they cover multiple pieces. If you want I’m happy to do a video call (and I can be flex on my timezone if that’s a constraint) and screen share so we can discuss the details.

**Comment by holdenk** [comment]
If it helps I called out the part of the PR I believe most likely responsible for that during my code review already.

**Comment by Ngone51** [comment]
> This PR in its original broke another one of the conditions.

Except for the breaking of the log(like you just fixed), what other conditions this PR breaks?

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:179]: First, we use `askSync` to send decommission notice to the driver whenever it needs(see `ExecutorSigPWRReceived`). Second, even if driver receives the decommission notice successfully, there still could be LaunchTask request due to the async between LaunchTask and decommission notice. Third, this part also uses async send,  so we still can not ensure the decommission notice is received by driver successfully.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:215]: What's race condition?

File: core/src/main/scala/org/apache/spark/deploy/DeployMessage.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/deploy/DeployMessage.scala:80]: yeah

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:74]: This's Worker, I guess you cares more about the executor? In Worker, `decommissionSelf` always returns true. and In exeutor, there's a change to return false to fail the decommissionSelf but seems rarely happen. If you would insist on returning the value, I think we can use askSync instead.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:1812]: ?

**Comment by Ngone51** [comment]
> That being said I still have concerns this PR is not sufficiently tested, can you add some more tests for the new flows you've introduced?

There's only one new flow that is from Master to Worker. I can update the existing test by verifying Worker's decommission status... What kind of other concerns do you have? Could you elaborate more? So I can improve the PR accordingly.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:179]: Right, so we should resend the notice then right?

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:215]: So we don’t ask the driver to stop scheduling jobs on us first, and the driver could ask us to run a job while we are part way through decommissioning. This won’t result in a failure because well accept the job but it will slow down the decommissioning. So swap the order of these two.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:1812]: Why did you make this change?

**Comment by holdenk** [comment]
I’m mostly concerned with the change around how the storage decommissioning is being done now, I’d like to see some tests that the flow from the master to the worker results in storage decommissioning.

**Comment by holdenk** [comment]
Also there are multiple log conditions, this PR broke one of them. Another PR had broken another one.

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:74]: Can you look into what the difference of this behavior might cause at the system level and then tell me if that’s a desired change? I’m ok with us making changes here, I just want us to be intentional and know if we need to test the change and it seems like this change was incidental.

**Comment by HyukjinKwon** [comment]
@Ngone51, shall we add a test https://github.com/apache/spark/pull/29817#issuecomment-698438316 and fix the test as requested? Seems like otherwise good to go.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:215]: Addressed

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:74]: The return value of the signal handling decides whether we should forward the signal to the other handlers. If `true`, no other handlers will handle the `PWR` signal except ourselves. If `false`, we will handle it (for decommission) and other handlers will handle it too.  Do you expect other handlers to continue handling the `SIGPWR` when the system isn't really experiencing a power failure?

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:179]: No. Sorry if I didn't explain it clearly.

We should already send decommission notice to the driver when `decommissioned = true`, which using `askSync`. If `askSync` still fail, I wouldn't expect another send would succeed.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:1813]: If I unsterstand your question correctly:

We didn't really change the `decommissionBlockManager`. The original `decommissionBlockManager` has been renamed to `decommissionSelf` to avoid the naming collision.

**Comment by Ngone51** [comment]
@holdenk Thanks for the review! I've addressed your most comments. And tests are updated and added in fa04b493324ea52698bed7ce15795530052b37be and
9d0f36d73e5750267ec2f33e7d964fc09778767f .

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by holdenk** [comment]
Tests are still failing @Ngone51 happy to take another look once the K8s integration tests are passing.

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:74]: I do, I think if the signal is unhandled then the process will be killed immediately. If we think of decommissioning/graceful shut down I believe that behavior is desirable, since if we can't shut down gracefully the least we can do is exit quickly.

**Review by holdenk — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:74]: Updated, but please note that I only updated for the executor's case since Worker's case always returns `true` before.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:473]: Not 100% sure, but I do remember that the locking mechanism withLock did manage to deadlock before, it might be good to add a logging statement in here that verifies this block is called in the integration test.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:501]: I think we might want to move this up because doing the storage decommissioning before the executor it's self is marked for decommissioning could mean that local blocks get stored in a manner where we might not know to migrate them. If you've dug through the code here and you don't believe that happens that's fine.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:1813]: Makes sense, although maybe introducing a new name instead of changing the use of a previous function name would be easier for verifying.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:166]: Maybe we can add a comment similar to the previous one of decommissionBlockManagers stating why we are doing this and it's intended effect.

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:682]: To make this code testable I'd move it down to inside of decommissionSelf and give decommissionSelf a param for if the decom originated locally.

File: core/src/test/scala/org/apache/spark/deploy/client/AppClientSuite.scala
Comment (holdenk) [inline core/src/test/scala/org/apache/spark/deploy/client/AppClientSuite.scala:127]: left a comment up above about how to improve the testing, I think that part of the test should cover the message being sent from the worker so we probably don't want to manually send that message here.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DecommissionSuite.scala
Comment (holdenk) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DecommissionSuite.scala:45]: nit: "to decommissioning." and "to decommissioning." are close enough you really only need the second to cover both cases since we're looking for the string anywhere. If you did want to you could switch this to regex and do more accurate matching. up to you.

**Review by holdenk — COMMENTED** [review]
Just a few notes, I noticed the integration tests are still failing. Let me know if you are stuck and would like some help figuring out what's going with the integration tests.

**Comment by Ngone51** [comment]
@holdenk Sorry for the delay. I had taken some days off recently. I'm addressing your comments right now and debugging the integration test(the decommissioned executor doesn't migrate shuffle blocks as expected, so I'm still investigating it.)

Thank you for the further review :)

**Comment by holdenk** [comment]
Taking days off is healthy and great! We're not in a rush on this so please take your time. My offer of help with debugging tests was not intended to make you feel like you needed to work on this right away, just I've been in the situation where I'm stuck on a test failure I can't figure out in the past. Let me know when you want me to take another look :)

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:682]: I added the param previously but removed it later in order to keep `decommissionSelf` unchanged according to the comment: https://github.com/apache/spark/pull/29722#discussion_r487855223

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:166]: Added.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:501]: We don't really migrate blocks when decommissioning BlockManagers above. We only mark them as being decommissioning at driver side. So I think the problem you mentioned won't exist.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DecommissionSuite.scala
Comment (Ngone51) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DecommissionSuite.scala]: Updated to use only one "decommissioning".

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:473]: > Not 100% sure, but I do remember that the locking mechanism withLock did manage to deadlock before

Yeah. We need to avoid the deadlock now because we need both CoarseGrainedSchedulerBackend's lock and TaskSchedulerImpl's lock within the same code block. 


> it might be good to add a logging statement in here that verifies this block is called in the integration test.

Updated the logs and the integration test.

**Comment by Ngone51** [comment]
cc @holdenk for taking another look. Thanks!

**Comment by Ngone51** [comment]
kindly ping @holdenk

**Comment by Ngone51** [comment]
@holdenk take a look?

**Comment by HyukjinKwon** [comment]
Let me merge this in few days if there are no more comments assuming lazy consensus.

**Comment by Ngone51** [comment]
@holdenk ?

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: core/src/main/scala/org/apache/spark/deploy/DeployMessage.scala
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/deploy/DeployMessage.scala:65]: shall we keep the param doc for `workerRef`?

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala:253]: unnecessary change?

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:179]: Yea I don't see the point of resending the notice to the driver, especially in this race condition. If we want to make sure the driver is noticed, we should design a mechanism for it, instead of doing it here randomly.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:218]: This part is new. I think it makes sense to skip executor decommissioning if the driver side rejects to do so.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:476]: I have a concern about the scope of the lock. Previously we only hold the lock when we change the mutable status, but now the entire method is locked. Do we have a good reason?

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:476]: Because now we not only change the mutable status but also call `scheduler.executorDecommission`. And `scheduler.executorDecommission` requires the lock on `TaskSchedulerImpl`. To avoid potential deadlock, we should use `withLock`.

More context: Previously,  change the mutable status and call `scheduler.executorDecommission` happens in two separate functions. In this PR, we combined these two functions into one function. Therefore, those two operations now are within the same code block and need the `withLock` protection.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by Ngone51** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by Ngone51** [comment]
thanks all!!

File: core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionIntegrationSuite.scala
Comment (MaxGekk) [inline core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionIntegrationSuite.scala:52]: I got test failure in my PR #31131 (the PR is not related to the test I believe):
```
[info] BlockManagerDecommissionIntegrationSuite:
[info] - SPARK-32850: BlockManager decommission should respect the configuration (enabled=false) *** FAILED *** (6 seconds, 165 milliseconds)
[info]   java.util.concurrent.TimeoutException: Can't find 2 executors before 6000 milliseconds elapsed
[info]   at org.apache.spark.TestUtils$.waitUntilExecutorsUp(TestUtils.scala:374)
[info]   at org.apache.spark.storage.BlockManagerDecommissionIntegrationSuite.$anonfun$new$2(BlockManagerDecommissionIntegrationSuite.scala:52)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
```

**Pull request closed without merging**
